### PR TITLE
[pool] Accept custom allocators

### PR DESCRIPTION
**Pull request by terrelln**
* pthreads functions might still allocate.

### CODE REVIEW

**Comment by Cyan4973** [comment]
Could be a good idea to retrofit `zstdmt` to use it

**Comment by terrelln** [comment]
Add `extern "C"` to `zstd_internal.h`.

File: lib/common/error_private.h [hunk 1]
```diff
@@ -51,7 +51,8 @@ typedef ZSTD_ErrorCode ERR_enum;
 #ifdef ERROR
 #  undef ERROR   /* reported already defined on VS 2015 (Rich Geldreich) */
```
Comment (Cyan4973) [inline lib/common/error_private.h]: very minor : 
it's allowed to `#undef` an identifier without previous definition.
So the `#ifdef` guard is not useful.

File: lib/common/pool.h [hunk 2]
```diff
@@ -16,6 +16,7 @@ extern "C" {
 
 
 #include <stddef.h>   /* size_t */
+#include "zstd_internal.h"   /* ZSTD_customMem */
```
Comment (Cyan4973) [inline lib/common/pool.h]: Are you comfortable with having `pool` depending on `zstd_internal` now ?

File: lib/common/threading.h [hunk 3]
```diff
@@ -37,7 +37,15 @@ extern "C" {
...
 #endif
 
+#ifdef ERROR
+#  undef ERROR   /* reported already defined on VS 2015 (Rich Geldreich) */
+#endif
 #include <windows.h>
+#ifdef ERROR
+#  undef ERROR   /* reported already defined on VS 2015 (Rich Geldreich) */
+#endif
+#define ERROR(name) ZSTD_ERROR(name)
```
Comment (Cyan4973) [inline lib/common/threading.h]: In which case is the `ERROR()` defined here useful ?

File: lib/common/pool.h (hunk 2 above)
Comment (terrelln) [reply lib/common/pool.h]: Its not a problem for me, if we ever need to extract it, we just need to delete the advanced versions.

File: lib/common/threading.h (hunk 3 above)
Comment (terrelln) [reply lib/common/threading.h]: Yeah, if `zstd_internal.h` is included after, then it isn't useful. But if it is included before, we've just undefined `ERROR`, so we need to redefine it.

**Pull request merged**